Fixing handling when pasted mutliple  file types

StaticBR opened the pull request:
New Microsoft Products also paste multiple Data transfer items, like:
![grafik](https://github.com/benwinding/quill-image-compress/assets/18309339/08e32b67-58d0-4cf3-b3ea-e3e615a6a127)

This PR  fixes this.

StaticBR commented:
Hi @benwinding ,
sure.
When copy /pasting from Current Excel or Outlook the TransferredDataItem contain multiple mime Types of data.
( See Screenshot) 
Currently the plugin does in this case always paste an image instead of the available HTML. 
Which is bad for the usability of the user.
So I changed the logic:  If there is any "text/html" in the TransferredDataItem[]  present, ignore the images and use the html.

benwinding commented:
What's different from the current behavior though? It seems to do exactly that, when pasting from other websites

Using the example site here -> https://benwinding.github.io/quill-image-compress/src/demo.html

https://github.com/benwinding/quill-image-compress/assets/11782590/5efaa357-e38a-4d33-a110-aee2e1fa00a3

StaticBR commented:
Hi Ben,
I'm so sorry there was an small error, I wanted to check the files not the images.
That's what i get for thinking I can fix a bug blindly.
Here you can see the difference when pasting from a local Excel :
![GIF 30 01 2024 11-20-58](https://github.com/benwinding/quill-image-compress/assets/18309339/adb9eab2-7b9c-40b4-a771-2275a82606f1)

Bests Ralf

StaticBR commented:
@benwinding: It would be be nice, if you can make a Bugfix Release for this issue. Thanks in advance.

StaticBR commented:
@benwinding  sry to bother you again.
Will you merge this change or do you have any concerns with it? 

I'm asking, because i have a currently pending Bug, in a Software depending on this. 
Thanks.

benwinding commented:
Hi StaticBR,

So it took me a while to understand this use case, but I think I've fixed it. Please try installing the new version and [try the updated demo here](https://benwinding.github.io/quill-image-compress/src/demo.html):

```
quill-image-compress@1.2.31
```

Does it work now? _(Apologies I can't merge your fix, as it broke the copy/paste in the browser)_

benwinding commented:
Hi @StaticBR, Thanks for adding more details 🙏  

> You said my solution "broke copy / paste in the browser", can you give me an example for that? Maybe I'm able to fix it.

Here's how your PR breaks copying an image (see video). 

User copies an image in the browser, which the following `fileTypes` in the `DataTransferItemList` from the clipboard

``` js
fileTypes: [
  {type: 'text/html', kind: 'string'},
  {type: 'image/png', kind: 'file'},
] 
```

Your PR ignores compressing the `image/png` due to it having `text/html` in the array. This is an extremely common use-case, which is why I can't merge this as it is...

https://github.com/benwinding/quill-image-compress/assets/11782590/a13a395f-d5fb-497c-b135-2228c8cf0de5

Are you able to record exactly what program you're copying from? (video is preferred over gif)

![image](https://github.com/benwinding/quill-image-compress/assets/11782590/a5e8c871-ff6d-406f-9837-7522d445377b)

Is this a cell in excel with the text "Critical"??? Don't know where this was copied from...

I'm on Mac and don't have any microsoft products to test... so if you can reproduce it with another website, or some mac compatible program that would be appreciated...

Would love to fix this, so give me as much detail as you can 🙏

StaticBR commented:
Hey Ben,

sry for the long waiting time. But your last post has pointed me to a few more issues.

If you currently paste an html code from a website the plugin will currently leave the images completely untouched. 
So it will basically keep the original img.src  of the image, even if this points to a remote URL.

I am not sure if that is whats people expecting. I surly was not.
By solving this issue I think I can also solve your and my issues together. 

So I added handling of pasted HTML to your plugin.
The html will now be parsed all images will be extracted and processed.

Here is an example of pasting html with an image from a foreign Url:
https://github.com/benwinding/quill-image-compress/assets/18309339/7ce73ca7-8562-4047-9ef7-901a1cd2000f

Even Pasting from Excel works now like a charm:
https://github.com/benwinding/quill-image-compress/assets/18309339/e3cd217f-2a0e-4e90-b746-408daec335a4

Unfortunately I had to add a **Breaking Change** to this pull request:
The option `insertIntoEditor` can no longer work if the plugin should be able to process html. Because of this I have replaced it with a  `uploadImage` method.  This allows us to also Upload images found within a pasted html. 

Please have a look at this change and let me know if you are willing to merge this change. 

Bests  Ralf

benwinding commented on src/quill.imageDrop.ts:
```diff
@@ -49,21 +46,24 @@ export class ImageDrop {
     if (evt.dataTransfer?.items) {
       this.logger.log("handleDrop", "found items", { evt, files, imageFiles });
-      await this.handleDataTransferList(evt.dataTransfer?.items, evt);
+      await this.handleDataTransferList(evt.dataTransfer!, evt);
```
`!` circumvents typescript's type-safety

benwinding commented on src/quill.imageDrop.ts:
```diff
@@ -72,15 +72,27 @@ export class ImageDrop {
 
-  private async handleDataTransferList(dataTransferItems: DataTransferItemList | undefined, evt: { preventDefault: () => void }) {
-    const items = Array.from(dataTransferItems || []);
+  private async handleDataTransferList(dataTransfer: DataTransfer, evt: Event) {
```
Probably need to change the method name to reflect the arguments a bit better

```suggestion
  private async handleDataTransfer(dataTransfer: DataTransfer | undefined, evt: Event) {
```

benwinding commented on src/quill.imageDrop.ts:
```diff
@@ -90,6 +102,27 @@ export class ImageDrop {
 
+  private async processHtml(html: string) {
+
+    const tmp = document.createElement('div')
+    tmp.innerHTML = html;
+    const imgNodeList = tmp.querySelectorAll('img');
+
+    for (let i = 0; i < imgNodeList.length; i++) {
+      const img = imgNodeList[i];
+      if (img.src.startsWith("http")) {
+        const blob = await (await fetch(img.src)).blob()
+        img.src = await file2b64(blob);
+      }
+      img.src = await this.processImage(img.src);
+    }
```
This is pretty clever! 🥇 However there could be `CORS`, or `401` errors during fetching, that would crash this whole `processHtml` function.

I think a good solution to these errors would be to add a placeholder image, instead of failing completely, WDYT?

Something like this maybe (encoded as a tiny base64 string and included in the library)

![](https://placehold.co/200?text=Image%20Unavailable&font=roboto)

benwinding commented on src/quill.imageDrop.ts:
```diff
@@ -72,15 +72,27 @@ export class ImageDrop {
+  private async handleDataTransferList(dataTransfer: DataTransfer, evt: Event) {
+    const items = Array.from(dataTransfer.items || []);
+
+    const html = dataTransfer.getData('text/html');
+    if (html) {
+      this.processHtml(html)
+        .catch(e => this.logger.error("error while processing pasted html", e));
+      evt.preventDefault();
+      return;
+    }
```
I like this, but would rather it be behind an option that people can opt-in/out of:

``` tsx
 if (html && options.allowPasteHTML)
```

Reason being, the default behavior of this PR calls `dangerouslyPasteHTML` which probably shouldn't be on by default due to [XSS security concerns](https://quilljs.com/docs/modules/clipboard/#dangerouslypastehtml)

StaticBR replied:
Used the name `compressImagesInPastedHtml` instead, to be more clear about what it does.

benwinding commented on src/quill.imageCompressor.ts:
```diff
@@ -100,23 +86,29 @@ class imageCompressor {
-      // Move cursor to next position
-      range.index++;
-      this.quill.setSelection(range, "api");
+  async processImage(dataUrl: string): Promise<string> {
+    dataUrl = await this.downscaleImageFromUrl(dataUrl);
+    if (this.options.uploadImage) {
+      dataUrl = await this.options.uploadImage(b64toBlob(dataUrl));
```
I find this confusing, it's modifying the `dataUrl` argument in several ways and also returning it 😕

benwinding commented on src/options.object.ts:
```diff
@@ -10,5 +10,5 @@ export type OptionsObject = {
   ignoreImageTypes?: string[];
   quality?: number;
-  insertIntoEditor?: (base64URL: string, imageBlob: Blob, editor: Quill) => void;
+  uploadImage?: (imageBlob: Blob) => Promise<string>;
```
Like you mentioned, this is a breaking change to the API, this would mean we'd need to release a new major version, which I don't think this change warrants.

At the very least, the current `insertIntoEditor` option should probably be maintained and marked as `@deprecated`

StaticBR replied:
I agree partiallay, but unfortunately we are not able to support this method when html is pasted, due to its method signature. 
My thought was to better have a breaking change than to have a legacy method which is only partially supported.

benwinding replied:
> but unfortunately we are not able to support this method when html is pasted, due to its method signature.

Can we at least support it for non-html pastes then? 

I'd really rather keep the API backwards compatible, as personally I hate it when libraries rename their API methods...

StaticBR replied:
Well I'm the complete opposite:  I rather like my code to fail fast instead of having an unexpected behavior that is lurking on spacial conditions. 

But it is your repository, so it is your decision to change it or not.

benwinding reviewed (commented):
Amazing efforts @StaticBR I really like what you've done here 👌
Will wait for your responses to my comments, I might clean it up before releasing it and it might require a "major version" bump due to the `insertIntoEditor` change, but hopefully we can figure a way around it 🙏

StaticBR commented:
FYI: I have added another Fix that prevented PNGs from retaining a transparent Background in some cases.
